### PR TITLE
Make applyProgram return Either instead of Maybe so that more error messages can be shown.

### PR DESCRIPTION
**Pull request by thealmarty**
Pre-submit checklist:
- Branch
    - [ ] Tests are provided (if possible)
    - [x] Commit sequence broadly makes sense
    - [x] Key commits have useful messages
    - [x] Changelog fragments have been written (if appropriate)
    - [x] Relevant tickets are mentioned in commit messages
    - [x] Formatting, PNG optimization, etc. are updated
- PR
    - [ ] (For external contributions) Corresponding issue exists and is linked in the description
    - [x] Targeting master unless this is a cherry-pick backport
    - [x] Self-reviewed the diff
    - [x] Useful pull request description
    - [x] Reviewer requested

### CODE REVIEW

File: plutus-core/executables/plc/Main.hs [hunk 1]
```diff
@@ -116,7 +115,9 @@ runApply (ApplyOptions inputfiles ifmt outp ofmt mode) = do
   let appliedScript =
         case map (\case p -> () <$ p) scripts of
           []          -> errorWithoutStackTrace "No input files"
-          progAndargs -> foldl1 (fromJust .* PLC.applyProgram) progAndargs
+          progAndargs ->
+            foldl1 (fromRight (error "applyProgram: cannot apply programs with different versions.")
```
Comment (thealmarty) [inline plutus-core/executables/plc/Main.hs:119]: Simulating `fromJust`, which errors out with `Nothing`.

Comment (zliu41) [inline plutus-core/executables/plc/Main.hs:119]: You can just show the error message from `PLC.applyProgram`

File: plutus-core/untyped-plutus-core/src/UntypedPlutusCore.hs [hunk 2]
```diff
@@ -25,7 +25,9 @@ applyProgram
     :: Semigroup a
     => Program name uni fun a
     -> Program name uni fun a
-    -> Maybe (Program name uni fun a)
+    -> Either String (Program name uni fun a)
```
Comment (zliu41) [inline plutus-core/untyped-plutus-core/src/UntypedPlutusCore.hs:28]: No String please. Make a proper error type.

File: plutus-tx/src/PlutusTx/Code.hs [hunk 3]
```diff
@@ -57,34 +57,51 @@ type CompiledCode = CompiledCodeIn PLC.DefaultUni PLC.DefaultFun
...
-    => CompiledCodeIn uni fun (a -> b) -> CompiledCodeIn uni fun a -> Maybe (CompiledCodeIn uni fun b)
+    :: (PLC.Closed uni
+        , uni `PLC.Everywhere` Flat
+        , Flat fun
+        , Pretty fun
+        , PLC.Everywhere uni PrettyConst
+        , PrettyBy RenderContext (PLC.SomeTypeIn uni))
+    => CompiledCodeIn uni fun (a -> b)
+    -> CompiledCodeIn uni fun a
+    -> Either String (CompiledCodeIn uni fun b)
```
Comment (thealmarty) [inline plutus-tx/src/PlutusTx/Code.hs:68]: I can add an error type for this too but I don't think it's beneficial in this case.

File: plutus-core/plutus-core/src/PlutusCore/Error.hs [hunk 4]
```diff
@@ -244,3 +245,12 @@ instance (tyname ~ TyName, name ~ Name) =>
...
 instance AsFreeVariableError (Error uni fun ann) where
     _FreeVariableError = _FreeVariableErrorE
+
+-- | Errors from `applyProgram` for PIR, PLC, UPLC.
+data ApplyProgramError =
+    MkApplyProgramError Version Version
+
+instance Show ApplyProgramError where
+    show (MkApplyProgramError v1 v2) =
+        "The first program has version " <> show v1
```
Comment (michaelpj) [inline plutus-core/plutus-core/src/PlutusCore/Error.hs:255]: include some context
```suggestion
        "Cannot apply two programs together: the first program has version " <> show v1
```

File: plutus-core/plutus-core/src/PlutusCore.hs [hunk 5]
```diff
@@ -148,7 +150,13 @@ applyProgram
...
-    -> Maybe (Program tyname name uni fun a)
+    -> Either ApplyProgramError (Program tyname name uni fun a)
 applyProgram (Program a1 v1 t1) (Program a2 v2 t2) | v1 == v2
-  = Just $ Program (a1 <> a2) v1 (Apply (termAnn t1 <> termAnn t2) t1 t2)
-applyProgram _ _ = Nothing
+  = Right $ Program (a1 <> a2) v1 (Apply (termAnn t1 <> termAnn t2) t1 t2)
+applyProgram (Program _a1 v1 _t1) (Program _a2 v2 _t2) =
+    Left $ MkApplyProgramError v1 v2
+
+getAppliedProgram :: Either ApplyProgramError prog -> prog
```
Comment (michaelpj) [inline plutus-core/plutus-core/src/PlutusCore.hs:159]: Not a huge fan of this, I think people can just deconstruct the either themselves how they want to. This is basically `either (error . show) id`

File: plutus-tx/src/PlutusTx/Code.hs (hunk 3 above)
Comment (michaelpj) [reply plutus-tx/src/PlutusTx/Code.hs:68]: Yeah, seems fine.

File: plutus-core/plutus-core/src/PlutusCore.hs (hunk 5 above)
Comment (michaelpj) [reply plutus-core/plutus-core/src/PlutusCore.hs:159]: I guess we do use it a bunch. I guess I'd rather have the more generalized
```
unsafeFromRight :: (Show e) => Either e a -> a
unsafeFromRight (Right a) = a
unsafeFromRight (Left e) = error $ show e
```
or something

Comment (thealmarty) [reply plutus-core/plutus-core/src/PlutusCore.hs:159]: If it's so general can I put it in the prelude instead?

**Comment by thealmarty** [comment]
Some cached failures on CI.

File: plutus-tx/src/PlutusTx/Code.hs [hunk 6]
```diff
@@ -12,20 +12,21 @@
...
-import PlutusCore qualified as PLC
-import PlutusIR qualified as PIR
-import PlutusTx.Coverage
-import PlutusTx.Lift.Instances ()
-import UntypedPlutusCore qualified as UPLC
-
 import Control.Exception
 import Data.ByteString qualified as BS
 import Data.ByteString.Lazy qualified as BSL
 import Data.Functor (void)
```
[the remark this answers is not in the document]
Comment (thealmarty) [reply plutus-tx/src/PlutusTx/Code.hs]: I see, it's now using `void` from the `PlutusPrelude`. Shouldn't be a problem.

[the remark this answers is not in the document]
Comment (michaelpj) [reply plutus-tx/src/PlutusTx/Code.hs]: You imported `PlutusPrelude` which re-exports `void`, hence this import is now redundant.

**Review by zliu41 — APPROVED** [review]

**Review by bezirg — APPROVED** [review]

**Pull request merged**